content/courses/advanced-ada/chapters/subprograms_modularity.rst

```diff
@@ -85,7 +96,53 @@ In this example, we declare :ada:`Calculate` as a private procedure of the
 :ada:`Data_Processing` package. Therefore, it's visible in that package (but
 not in the :ada:`Test_Data_Processing` procedure).
 
-.. admonition:: In the Ada Reference Manual
 
-    - :arm:`10.1.1 Compilation Units - Library Units <10-1-1>`
-    - :arm:`10.1.2 Context Clauses - With Clauses <10-1-2>`
+Private subprograms and private packages
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+We can also use private subprograms to test private packages. As we know, in
+most cases, we cannot access private packages in external clients |mdash| such
+as external subprograms. However, by declaring a subprogram private, we're
+allowed to access private packages. This can be very useful to create
+applications that we can use to test private packages:
+
+.. code:: ada run_button main=test_private_data_processing.adb project=Courses.Advanced_Ada.Subprograms.Private_Subprogram_Private_Package
+
+    private package Private_Data_Processing is
+
+       type Data is private;
+
+       procedure Process (D : in out Data);
+
+    private
+
+        type Data is null record;
+
+    end Private_Data_Processing;
+
+    package body Private_Data_Processing is
+
+       procedure Process (D : in out Data) is
+       begin
+          null;
+       end Process;
+
+    end Private_Data_Processing;
+
+    private procedure Test_Private_Data_Processing;
```

Comment:
Let me also suggest you show making the private procedure be a child of the private package you want to test. That would simply things a bit, but more importantly would allow refereeing the private part of the package, where data and or hidden subprograms would be declared, thereby creating a more extensive test in some cases.

Reply:
Yes, good point! I extended the section and added a code example for that.